crates/syntax/src/ast/node_ext.rs

```diff
@@ -297,6 +297,52 @@ impl ast::RecordExprField {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum NameLike {
+    NameRef(ast::NameRef),
+    Name(ast::Name),
+    Lifetime(ast::Lifetime),
+}
+
+impl NameLike {
+    pub fn as_name_ref(&self) -> Option<&ast::NameRef> {
+        match self {
+            NameLike::NameRef(name_ref) => Some(name_ref),
+            _ => None,
+        }
+    }
+}
+
+impl ast::AstNode for NameLike {
+    fn can_cast(kind: SyntaxKind) -> bool {
+        matches!(kind, SyntaxKind::NAME | SyntaxKind::NAME_REF | SyntaxKind::LIFETIME)
+    }
+    fn cast(syntax: SyntaxNode) -> Option<Self> {
+        let res = match syntax.kind() {
+            SyntaxKind::NAME => NameLike::Name(ast::Name { syntax }),
+            SyntaxKind::NAME_REF => NameLike::NameRef(ast::NameRef { syntax }),
+            SyntaxKind::LIFETIME => NameLike::Lifetime(ast::Lifetime { syntax }),
+            _ => return None,
+        };
+        Some(res)
+    }
+    fn syntax(&self) -> &SyntaxNode {
+        match self {
+            NameLike::NameRef(it) => it.syntax(),
+            NameLike::Name(it) => it.syntax(),
+            NameLike::Lifetime(it) => it.syntax(),
+        }
+    }
+}
+
+mod __ {
+    use super::{
+        ast::{Lifetime, Name, NameRef},
+        NameLike,
+    };
+    stdx::impl_from!(NameRef, Name, Lifetime for NameLike);
+}
+
 #[derive(Debug, Clone, PartialEq)]
 pub enum NameOrNameRef {
```

Comment:
Hm, are there usages of NameOrNameRef which should be replaced by NameLike?

Reply:
The only thing that uses `NameOrNameRef` is `ast::RecordPatField::field_name`, replacing it with NameLike would look like this [`e1dbf43` (#7699)](https://github.com/rust-analyzer/rust-analyzer/pull/7699/commits/e1dbf43cf85f84c3a7e40f9731fc1f7ac96f8979), assuming keeping both isn't worth it.

Reply:
ah right, field shorthand is where we need exactly this...

bors r+

Reply:
oh, wait, this is with them merged, lets keep the split bors r-